specs/capella/beacon-chain.md

```diff
@@ -371,15 +272,29 @@ def process_block(state: BeaconState, block: BeaconBlock) -> None:
 
 ```python
 def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
-    num_withdrawals = min(MAX_WITHDRAWALS_PER_PAYLOAD, len(state.withdrawal_queue))
-    dequeued_withdrawals = state.withdrawal_queue[:num_withdrawals]
-
-    assert len(dequeued_withdrawals) == len(payload.withdrawals)
-    for dequeued_withdrawal, withdrawal in zip(dequeued_withdrawals, payload.withdrawals):
-        assert dequeued_withdrawal == withdrawal
-
-    # Remove dequeued withdrawals from state
-    state.withdrawal_queue = state.withdrawal_queue[num_withdrawals:]
+    epoch = get_current_epoch(state)
+    index = state.last_withdrawal_validator_index
+    for withdrawal in payload.withdrawals:
```

Comment:
Rather than iterating through payloads given in the block, I think using the same method as in block production would be smart (`get_expected_withdrawals`). To make sure for a given state, the correct set of withdrawals is chosen, and we can just check equivalence. It seems like it'd be less error prone if we used it block production and processing

Reply:
currently `get_expected_withdrawals` is specified in terms of the withdrawal queue, so it would need to be re-phrased, likely as an iteration over the validator set (which clients then need to optimize out with a cache).

Reply:
Yeah I want to explicitly leave this out of the spec and let clients do whatever they want to optimize this scan. Unless the network is leaking for 5 straight days, it should be trivial to find the next 16 withrawals. Anyway a circular buffer cache of fixed size would most probably be our implementation if we go with this. I'll clean the PR up this week. Many thanks for the reviews

Reply:
Awesome.

I think it doesn't hurt to also define the semantics of `get_expected_withdrawals` in an inefficient way in the spec. This would likely be required for generating and executing spec tests anyway.